hypergpufunctions.cu

```diff
@@ -140,58 +189,77 @@ void HyperFunctionsGPU::spec_sim_GPU() {
+    //spec_simil_img=test_img1;
+    this->oneD_array_to_mat(out, mlt1[1].rows, mlt1[1].cols);   
 }
+
 void HyperFunctionsGPU::deallocate_memory() 
 {
     cudaFree(d_img_array); cudaFree(d_ref_spectrum); cudaFree(d_out);
 }
 
-
-
+/* allocating CUDA memory. 
+* In cuda, grids contain blocks of threads, which are used for parallel computations. 
+* One pixel in img_array needs one thread for a computation, so threads = number of pixels. 
+* Set the number of threads per block to be 512 - this is the maximum threads per block for older GPUs.
+* Grid size is the number of of blocks, given by the number of threads / block size + 1. 
+*/
 void HyperFunctionsGPU::allocate_memory(int* img_array) {
-    //allocating CUDA memory
+    
     N_points=mlt1[1].rows*mlt1[1].cols*mlt1.size(); 
     N_size=mlt1[1].rows*mlt1[1].cols;    
     num_lay=  mlt1.size();
     block_size = 512;
-    grid_size = ((N_points + block_size) / block_size);
-
+    grid_size = ((N_points + block_size) / block_size); 
+    
     int tmp_len1=reference_spectrums[ref_spec_index].size(); 
 
-    cudaHostAlloc ((void**)&out, sizeof(int) *N_size, cudaHostAllocDefault);
+    cudaHostAlloc ((void**)&out, sizeof(unsigned int) * N_size, cudaHostAllocDefault);
     cudaMalloc((void**)&d_img_array, sizeof(int) * N_points);
-    cudaMalloc((void**)&d_out, sizeof(int) * N_size);
+    cudaMalloc((void**)&d_out, sizeof(unsigned int) * N_size);
     cudaMalloc((void**)&d_ref_spectrum, sizeof(int) * tmp_len1);
 
+    //allocating memory on the GPU device
+
     int* ref_spectrum=new int[tmp_len1];
     for (int i=0;i<reference_spectrums[ref_spec_index].size();i++)
     {
-        ref_spectrum[i]=reference_spectrums[ref_spec_index][i]  ;
+        ref_spectrum[i] = reference_spectrums[ref_spec_index][i]; 
+        //converting the 2-D array of reference spectrums into one-D for CUDA processing
     }
     cudaMemcpy(d_img_array, img_array, sizeof(int) *  N_points, cudaMemcpyHostToDevice);
     cudaMemcpy(d_ref_spectrum, ref_spectrum, sizeof(int) * tmp_len1, cudaMemcpyHostToDevice);
-}
 
+    //copying our existing memory that holds image data and reference spectrum data to the 
+    //allocated memory on the GPU
+}
 
-void HyperFunctionsGPU::oneD_array_to_mat(int* img_array )
+/**
+ * Converts one-D array to 1 channel 8 bit OPENCV matrix. 
+ * Used in manipulating spectral similarity data. 
+*/
+void HyperFunctionsGPU::oneD_array_to_mat(int* img_array, int rows, int cols)
 {
-
-    int val_it=0;
+    /*int val_it=0;
     for (int k=0; k<spec_simil_img.cols; k++)
     {
         for (int j=0; j<spec_simil_img.rows; j++)
         {
-            spec_simil_img.at<uchar>(j,k)=img_array[val_it];
+            spec_simil_img.at<uchar>(j,k)=img_array[val_it]; 
             val_it+=1;
+            //Conversion of 1-d array containing pixel values to 8 bit one channel 2-d OPENCV matrix
         }
-    }
+    }*/
+    spec_simil_img = cv::Mat(rows, cols, CV_32SC1, img_array);
```

Comment:
can we use mlt1[0].rows / cols to figure out rows and cols, so we do not need to pass the values?

Reply:
spec_simil_img.convertTo(spec_simil_img, CV_8UC1);
conversion should also be done here

Reply:
fixed